### PR TITLE
Feature: Implement tokenURI for Staking and Cover contracts

### LINKED ISSUE #372: Implement tokenURI() for Cover and StakingNFTs
Options;
1.) We can implement a link to an ipfs address
2.) Let the tokenURI() function build and encode the data we intend to store

### PR DESCRIPTION
**Pull request by kyledewy**
## Context

Closes #372 

## Changes proposed in this pull request

Implemented `tokenUri(id)` to return a json object with information about that NFT. The NFTs are not upgradeable so most of the heavy lifting of the tokenURI() function is done in the `Descriptor`  contracts. 
The SVG is just a placeholder and still needs to be designed. 

## Test plan

There are some tests that the json is structured properly and that the tokens handle expiration/non-existant queries properly. 

## Checklist

- [x] Rebased the base branch
- [x] Attached corresponding Github issue
- [x] Prefixed the name with the type of change (i.e. feat, chore, test)
- [x] Performed a self-review of my own code
- [x] Followed the style guidelines of this project
- [x] Made corresponding changes to the documentation
- [x] Didn't generate new warnings
- [x] Didn't generate failures on existing tests
- [x] Added tests that prove my fix is effective or that my feature works

## Review

When reviewing a PR, please indicate intention in comments using the following emojis:
* :cake: = Nice to have but not essential.
* :bulb: = Suggestion or a comment based on personal opinion
* :hammer: = I believe this should be changed.
* :thinking: = I don’t understand something, do you mind giving me more context?
* :rocket: = Feedback

### CODE REVIEW

File: contracts/modules/cover/CoverNFTDescriptor.sol [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+import "./CoverNFT.sol";
+
+contract CoverNFTDescriptor {
+  using Strings for uint256;
+  using DateTime for uint256;
+
+  string[] public coverAssets;
+
+  constructor() {
+    // TODO: how to handle asset symbols?
```
Comment (kyledewy) [inline contracts/modules/cover/CoverNFTDescriptor.sol:17]: Not sure what to do about displaying the symbols. Do we expect new cover assets often?

File: contracts/modules/cover/CoverNFT.sol [hunk 2]
```diff
@@ -3,33 +3,43 @@
...
+    CoverTokenURIParams memory params = CoverTokenURIParams(id, name);
+    uri = ICoverNFTDescriptor(nftDescriptor).tokenURI(params);
   }
 
   function isApprovedOrOwner(address spender, uint tokenId) external view returns (bool) {
     address owner = ownerOf(tokenId);
     return spender == owner || isApprovedForAll[owner][spender] || spender == getApproved[tokenId];
   }
 
+  function changeNFTDescriptor(address _newNFTDescriptor) public onlyOperator returns (bool) {
```
Comment (kyledewy) [inline contracts/modules/cover/CoverNFT.sol:36]: Not sure who should have authority to call this function.

File: contracts/modules/staking/StakingNFTDescriptor.sol [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+      uint newRewardPerShare = stakingPool.getAccNxmPerRewardsShare() - lastAccNxmPerRewardShare;
+      pendingRewards += _pendingRewards + (newRewardPerShare * _rewardsShares) / ONE_NXM;
+
+      // update total stake
+      uint stake = (activeStake * _stakeShares) / stakeSharesSupply;
+      totalStake += stake;
+
+      // calculate days left until stake expires
+      uint daysLeftInTranche = (TRANCHE_DURATION - (block.timestamp % TRANCHE_DURATION)) / 1 days;
+      uint daysLeft = daysLeftInTranche + (i * TRANCHE_DURATION / 1 days);
```
Comment (kyledewy) [inline contracts/modules/staking/StakingNFTDescriptor.sol:91]: Should I turn this into DD/MM/YYYY format?

File: contracts/modules/cover/CoverNFT.sol [hunk 4]
```diff
@@ -85,6 +87,13 @@ contract CoverNFT is ICoverNFT {
     return _balanceOf[owner];
   }
 
+  function changeNFTDescriptor(address _newNFTDescriptor) public onlyOperator returns (bool) {
```
Comment (shark0der) [inline contracts/modules/cover/CoverNFT.sol:90]: Remove return variable

File: contracts/modules/cover/CoverNFTDescriptor.sol (hunk 1 above)
Comment (shark0der) [reply contracts/modules/cover/CoverNFTDescriptor.sol:18]: Passing them via constructor would force us to deploy a new token descriptor every time the assets change. Should be pulled from the Pool contract instead. Probably could pass the master address via constructor and make it immutable.

File: contracts/modules/cover/CoverNFTDescriptor.sol [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+import "@openzeppelin/contracts-v4/utils/Strings.sol";
+import "@openzeppelin/contracts-v4/utils/Base64.sol";
+import "../../interfaces/ICover.sol";
+import "../../interfaces/ICoverNFTDescriptor.sol";
+import "../../libraries/DateTime.sol";
+import "./CoverNFT.sol";
+
+contract CoverNFTDescriptor {
+  using Strings for uint256;
+  using DateTime for uint256;
```
Comment (shark0der) [inline contracts/modules/cover/CoverNFTDescriptor.sol:13]: Use `uint` instead for consistency with the rest of the project.

Comment (shark0der) [reply contracts/modules/cover/CoverNFTDescriptor.sol:13]: ^ should be applied on the whole PR.

File: contracts/interfaces/ICoverNFTDescriptor.sol [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+pragma solidity >=0.5.0;
+
+  struct CoverTokenURIParams {
+    uint256 tokenId;
+    string name;
+  }
+
+interface ICoverNFTDescriptor {
+
+  function tokenURI(CoverTokenURIParams calldata params) external view returns (string memory);
```
Comment (shark0der) [inline contracts/interfaces/ICoverNFTDescriptor.sol:12]: Let's avoid the struct, pass the 2 variables directly.

File: contracts/interfaces/IStakingNFTDescriptor.sol [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    uint256 tokenId;
+    uint256 poolId;
+    address stakingPool;
+    address owner;
+    string name;
+  }
+
+interface IStakingNFTDescriptor {
+
+  function tokenURI(StakingTokenURIParams calldata params) external view returns (string memory);
```
Comment (shark0der) [inline contracts/interfaces/IStakingNFTDescriptor.sol:20]: Let's avoid the struct, pass the variables directly.

File: contracts/modules/cover/CoverNFTDescriptor.sol [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+          )
+        )
+      )
+    );
+  }
+
+  function generateDescription(CoverTokenURIParams calldata params) public view returns (string memory) {
+
+    CoverNFT coverNFT = CoverNFT(msg.sender);
+    ICover cover = ICover(coverNFT.operator());
```
Comment (shark0der) [inline contracts/modules/cover/CoverNFTDescriptor.sol:48]: Fetch the cover address from master instead, as the operator is not guaranteed to be the Cover contract in the future.

File: contracts/modules/cover/CoverNFTDescriptor.sol [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+    // Get cover data
+    CoverData memory coverData = cover.coverData(params.tokenId);
+    string memory productName = cover.productNames(coverData.productId);
+    CoverSegment memory lastSegment = cover.coverSegments(params.tokenId, cover.coverSegmentsCount(params.tokenId) - 1);
+
+    // Get expiration timestamp and check if expired
+    uint expiryTimestamp = lastSegment.start + lastSegment.period;
+
+    // Format date to DD/MM/YYYY
```
Comment (shark0der) [inline contracts/modules/cover/CoverNFTDescriptor.sol:62]: image

File: contracts/modules/cover/CoverNFTDescriptor.sol [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+        " \\nExpiry Date: ", expiry, " \\n", expiryMessage
+      )
+    );
+    return description;
+  }
+
+  function generateSVGImage(CoverTokenURIParams calldata params) public pure returns (bytes memory) {
+    return abi.encodePacked(
+      string(
+        abi.encodePacked(
```
Comment (shark0der) [inline contracts/modules/cover/CoverNFTDescriptor.sol:88]: can we haz preview? :grin:

File: contracts/modules/staking/StakingNFT.sol [hunk 11]
```diff
@@ -50,7 +54,10 @@ contract StakingNFT is IStakingNFT {
     operator = newOperator;
   }
 
-  // TODO: implement change token descriptor function here
+  function changeNFTDescriptor(address newNFTDescriptor) public {
+    require(msg.sender == operator, "NOT_OPERATOR");
```
Comment (shark0der) [inline contracts/modules/staking/StakingNFT.sol:58]: Custom error

File: contracts/modules/staking/StakingNFTDescriptor.sol (hunk 3 above)
Comment (shark0der) [reply contracts/modules/staking/StakingNFTDescriptor.sol:91]: Should probably show it as expiration date, yes. Easier to grasp.

File: contracts/interfaces/IStakingNFTDescriptor.sol (hunk 7 above)
Comment (kyledewy) [reply contracts/interfaces/IStakingNFTDescriptor.sol:20]: stack is pretty deep. whats the reason to pass them directly?

File: contracts/modules/cover/CoverNFTDescriptor.sol (hunk 9 above)
Comment (kyledewy) [reply contracts/modules/cover/CoverNFTDescriptor.sol:62]: I went with `Sep 3 2023` format for readability. It's similar to the javascript datetime library `toDateString()`, just missing the day of the week. I think the day of the week is overkill though.

Comment (roxdanila) [reply contracts/modules/cover/CoverNFTDescriptor.sol]: yeah, sounds good 👍

**Comment by roxdanila** [comment]
@kyledewy let's rebase, fix tests etc. 
given the timeframe, let's keep things simple on the staking side and only include the PoolId for now. we can always revisit later.

I'm also waiting for the SVGs from the designer, they should be ready shortly.

**Comment by kyledewy** [comment]
The Cover amount is overlapping :thinking: 
overlap

**Comment by kyledewy** [comment]
Current SVGs: 

NonExistingCover
image

Expired/ActiveCover
image

Stake
image

Expired Stake
image

NonExistingStake 
image

File: contracts/libraries/FloatingPoint.sol [hunk 12]
```diff
@@ -0,0 +1,52 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+import "@openzeppelin/contracts-v4/utils/Strings.sol";
+
+pragma solidity ^0.8.0;
+
+contract FloatingPoint {
```
Comment (kyledewy) [inline contracts/libraries/FloatingPoint.sol:6]: I was having issues linking this as a library for some reason. I tried inlining OZ dependency but that didn't help.

Comment (roxdanila) [reply contracts/libraries/FloatingPoint.sol:6]: you had a public function, looks like this was the reason

**Review by roxdanila — APPROVED** [review]

**Review by shark0der — APPROVED** [review]

**Pull request merged**